### PR TITLE
fix(components/datetime): date range picker visually invalid on custom errors

### PR DESCRIPTION
**Pull request by Blackbaud-SandhyaRajasabeson**
[#AB3140330](https://dev.azure.com/blackbaud/Products/_boards/board/t/SKY%20UX%20Program/Stories/?workitem=3140330)

### CODE REVIEW

**Comment by blackbaud-sky-build-user** [comment]
[Storybook preview](https://blackbaud.github.io/skyux-pr-preview/2910/storybook/)

Component Storybooks:

- [ag-grid-storybook](https://blackbaud.github.io/skyux-pr-preview/2910/storybooks/ag-grid-storybook/)
- [datetime-storybook](https://blackbaud.github.io/skyux-pr-preview/2910/storybooks/datetime-storybook/)
- [lists-storybook](https://blackbaud.github.io/skyux-pr-preview/2910/storybooks/lists-storybook/)
- [tabs-storybook](https://blackbaud.github.io/skyux-pr-preview/2910/storybooks/tabs-storybook/)

Apps:

- [code-examples](https://blackbaud.github.io/skyux-pr-preview/2910/code-examples/)
- [integration](https://blackbaud.github.io/skyux-pr-preview/2910/integration/)
- [playground](https://blackbaud.github.io/skyux-pr-preview/2910/playground/)

File: apps/code-examples/src/app/code-examples/datetime/date-range-picker/basic/demo.component.ts [hunk 1]
```diff
@@ -33,7 +34,12 @@ function dateRangeExcludesWeekend(
   standalone: true,
   selector: 'app-demo',
   templateUrl: './demo.component.html',
-  imports: [FormsModule, ReactiveFormsModule, SkyDateRangePickerModule],
+  imports: [
+    FormsModule,
+    ReactiveFormsModule,
+    SkyDateRangePickerModule,
+    CommonModule,
```
Comment (Blackbaud-SteveBrush) [inline apps/code-examples/src/app/code-examples/datetime/date-range-picker/basic/demo.component.ts]: Was the code example not working without `CommonModule`?

File: libs/components/datetime/src/lib/modules/date-range-picker/date-range-picker.component.spec.ts [hunk 2]
```diff
@@ -876,6 +876,69 @@ describe('Date range picker', function () {
...
+
+    expect(
+      startInput.classList.contains('sky-field-status-invalid'),
+    ).toBeTrue();
+    expect(
+      calculatorInput.classList.contains('sky-field-status-invalid'),
+    ).toBeTrue();
+  }));
+
+  it('should display custom form errors and mark mark all inputs as invalid', fakeAsync(() => {
```
Comment (Blackbaud-SteveBrush) [inline libs/components/datetime/src/lib/modules/date-range-picker/date-range-picker.component.spec.ts]: "mark mark"

File: libs/components/datetime/src/lib/modules/date-range-picker/date-range-picker.component.spec.ts [hunk 3]
```diff
@@ -876,6 +876,69 @@ describe('Date range picker', function () {
     });
   });
 
+  it('should mark select calculator picker as invalid only when the entire date range picker is invalid', fakeAsync(() => {
```
Comment (Blackbaud-SteveBrush) [inline libs/components/datetime/src/lib/modules/date-range-picker/date-range-picker.component.spec.ts]: I'm having difficulty understanding this statement. If we change it to, "should only mark the calculator select as invalid if the range is invalid" does it still communicate what the test is capturing?

File: libs/components/datetime/src/lib/modules/date-range-picker/date-range-picker.component.ts [hunk 4]
```diff
@@ -554,6 +563,25 @@ export class SkyDateRangePickerComponent
...
   }
 
+  #createHostCustomErrorChangeSignal(
+    control: AbstractControl,
+  ): Signal<boolean | undefined> {
+    return toSignal(
+      control.events.pipe(
+        filter((evt) => evt instanceof StatusChangeEvent),
+        map((evt: StatusChangeEvent) => {
+          const errors: ValidationErrors | null = evt.source.errors;
```
Comment (Blackbaud-SteveBrush) [inline libs/components/datetime/src/lib/modules/date-range-picker/date-range-picker.component.ts]: Just a few minor things:
- The value of `errors` can be inferred from the assignment.
- Adding the known errors to an array will 1) self-document the purpose of the function, and 2) allow future engineers an easy way to add more errors to the list
```
const errors = evt.source.errors ?? [];
const knownErrors = ['required', 'skyDate'];

return Object.keys(errors).some((error) => {
  return !knownErrors.includes(error);
});
```

File: apps/code-examples/src/app/code-examples/datetime/date-range-picker/basic/demo.component.ts (hunk 1 above)
Comment (Blackbaud-SandhyaRajasabeson) [reply apps/code-examples/src/app/code-examples/datetime/date-range-picker/basic/demo.component.ts]: oop that was a miss from testing something else and i needed the json pipe lol

File: libs/components/datetime/src/lib/modules/date-range-picker/date-range-picker.component.spec.ts (hunk 3 above)
Comment (Blackbaud-SandhyaRajasabeson) [reply libs/components/datetime/src/lib/modules/date-range-picker/date-range-picker.component.spec.ts]: kinda? calculator is invalid if the range is invalid and not for any datepicker specific errors. but its also invalid on other errors that have to do with the entire date range picker like custom errors. i agree that the wording is not clear tho. lemme think on it

File: libs/components/datetime/src/lib/modules/date-range-picker/date-range-picker.component.ts (hunk 4 above)
Comment (Blackbaud-SandhyaRajasabeson) [reply libs/components/datetime/src/lib/modules/date-range-picker/date-range-picker.component.ts]: oh thats much cleaner 🫡

File: libs/components/datetime/src/lib/modules/date-range-picker/date-range-picker.component.spec.ts (hunk 3 above)
Comment (Blackbaud-SandhyaRajasabeson) [reply libs/components/datetime/src/lib/modules/date-range-picker/date-range-picker.component.spec.ts]: `should not mark calculator as invalid when the error is from the datepicker` ?

Comment (Blackbaud-SteveBrush) [reply libs/components/datetime/src/lib/modules/date-range-picker/date-range-picker.component.spec.ts]: That works.

**Review by Blackbaud-SteveBrush — APPROVED** [review]

**Pull request merged**